src/main/scala/org/apache/spark/metrics/source/custom/AccumulatorMetricsSource.scala

```diff
@@ -0,0 +1,94 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.apache.spark.metrics.source.custom
+// This class is defined under `org.apache.spark` package, because it extends
+// package-private `org.apache.spark.metrics.source.Source`
+
+import java.util.concurrent.{Executors, TimeUnit}
+import java.{lang => jl}
+
+import com.codahale.metrics.MetricRegistry
+import org.apache.commons.lang3.concurrent.BasicThreadFactory
+import org.apache.log4j.LogManager
+import org.apache.spark.SparkEnv
+import org.apache.spark.metrics.source.Source
+import org.apache.spark.util.{AccumulatorV2, ShutdownHookManager}
+
+import scala.collection.mutable
+
+/**
+  * Accumulator-based metric source.
+  *
+  * This source uses Spark accumulators for collecting metrics. After registering an accumulator, it will be
+  * periodically polled for its value which will be reported to Spark metrics system using [[com.codahale.metrics.Meter]].
+  * After accumulators are registered it is necessary to start the source with [[AccumulatorMetricsSource#start()]],
+  * which will start internal polling thread and register shutdown hook for flushing metrics.
+  *
+  * Usage:
+  * {{{
+  * val metricsSource = new AccumulatorMetricsSource("prefix")
+  * val metric = spark.sparkContext.longAccumulator("metric-name")
+  * metricsSource.registerAccumulator(metric)
+  * metricsSource.start()
+  * }}}
+  *
+  * @param prefix for metrics collected by this source
+  */
+class AccumulatorMetricsSource(prefix: String) extends Source {
+  override val sourceName: String = prefix
+  override val metricRegistry: MetricRegistry = new MetricRegistry
+
+  private val log = LogManager.getLogger(this.getClass.getName)
+
+  private val accumulators = mutable.Map[String, AccumulatorV2[jl.Long, jl.Long]]()
+
+  private val executor = Executors.newSingleThreadScheduledExecutor(
+    new BasicThreadFactory.Builder()
+      .namingPattern("accumulator-metrics-source-thread-%d")
+      .daemon(true)
+      .priority(Thread.NORM_PRIORITY)
+      .build())
+
+  def registerAccumulator(acc: AccumulatorV2[jl.Long, jl.Long]) {
+    val name = acc.name.getOrElse("UNNAMED_ACCUMULATOR")
+    require(!accumulators.keySet.contains(name), "Accumulator already registered")
+    accumulators.put(name, acc)
+    metricRegistry.meter(name)
```

Comment:
This method should define return type `: Unit`.

Also,  is this call to `metricRegistry.meter(name)` necessary? Do we want to make sure the meter is created at this point? Since we don't use the return value, a comment about intention would be helpful.

Reply:
> This method should define return type : Unit.

:+1: 

> Also, is this call to metricRegistry.meter(name) necessary? Do we want to make sure the meter is created at this point? Since we don't use the return value, a comment about intention would be helpful.

Adding a comment here, it's better to keep it in order to catch registration errors as early as possible.